src/Workspaces/Core/Portable/Workspace/Solution/Project.cs

```diff
@@ -239,19 +242,37 @@ public bool ContainsAnalyzerConfigDocument(DocumentId documentId)
     /// Get the document in this project with the specified document Id.
     /// </summary>
     public Document? GetDocument(DocumentId documentId)
-        => ImmutableInterlocked.GetOrAdd(ref _idToDocumentMap, documentId, s_tryCreateDocumentFunction, this);
+    {
+        lock (_gate)
+        {
+            _idToDocumentMap ??= [];
+            return _idToDocumentMap.GetOrAdd(documentId, s_tryCreateDocumentFunction, this);
+        }
+    }
 
     /// <summary>
     /// Get the additional document in this project with the specified document Id.
     /// </summary>
     public TextDocument? GetAdditionalDocument(DocumentId documentId)
-        => ImmutableInterlocked.GetOrAdd(ref _idToAdditionalDocumentMap, documentId, s_tryCreateAdditionalDocumentFunction, this);
+    {
+        lock (_gate)
+        {
+            _idToAdditionalDocumentMap ??= [];
+            return _idToAdditionalDocumentMap.GetOrAdd(documentId, s_tryCreateAdditionalDocumentFunction, this);
+        }
+    }
 
     /// <summary>
     /// Get the analyzer config document in this project with the specified document Id.
     /// </summary>
     public AnalyzerConfigDocument? GetAnalyzerConfigDocument(DocumentId documentId)
-        => ImmutableInterlocked.GetOrAdd(ref _idToAnalyzerConfigDocumentMap, documentId, s_tryCreateAnalyzerConfigDocumentFunction, this);
+    {
+        lock (_gate)
+        {
+            _idToAnalyzerConfigDocumentMap ??= [];
+            return _idToAnalyzerConfigDocumentMap.GetOrAdd(documentId, s_tryCreateAnalyzerConfigDocumentFunction, this);
+        }
+    }
```

Comment:
small nit.  consider extracting helper for this pattern.  we often have triplicate methods for docs/analyzerdocs/additionaldocs.  i lke it when they all call into a single helper.

Reply:
Extracted both GetOrAdd and TryGetValue calls into their own methods that can be called for any of the dictionaries.